### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by arelaxend**
Hi! HNY!

I have finished a complete rework of the project because there were several issues related to:
- **Optimization**: we must restart with a fresh and highly optimize solution. The current solution is fast, and linear in time, and the distribution of document has a equiprobable distribution at infinity,
- **bugs (free)**: some bugs have been solved,
- **simplification**,
- **scala friendly**: several years ago, I'd got 100% at the course of Prof. Martin Odersky on Scala (not the online course!!) ;)

A.

### CODE REVIEW

**Comment by thammegowda** [comment]
Hi @arelaxend HNY to you too..

this is great.
I appreciate your time spent to make these changes.
Thanks for starting to optimize, simplify and to fix bugs.

I will give you comments in the code review and ask you more details when required.

as of now, it would be great to have a brief description of bugs, optimizations, simplifications etc (for the sake of record keeping and informing the team)

Thanks :-) 
T.G.

File: sparkler-app/src/main/scala/edu/usc/irds/sparkler/service/SolrProxy.scala [hunk 1]
```diff
@@ -20,41 +20,38 @@ package edu.usc.irds.sparkler.service
...
+    for (doc: Resource <- beans) {
+      if (crawlDb.getById(doc.id) == null) {
+        try {
+          crawlDb.addBean(doc)
+        } catch {
+          case e2: Exception =>
+            LOG.debug(e2.getMessage, e2)
         }
+      }
     }
```
Comment (thammegowda) [inline sparkler-app/src/main/scala/edu/usc/irds/sparkler/service/SolrProxy.scala]: The proposed changes feels simpler but inefficient.

`addBeans()` does batch insertion task.
`addBean()` inserts one document at a time.

So, if we want to insert 1000 docs, the proposed code does 1000 different insertions.

The old strategy was to attempt batch insertion, on failure fall back to individual insertion by skipping+logging the erroneous docs. Having erroneous docs (such with encoding issues) was rare and thus it was mostly a batch insertion.

**Comment by thammegowda** [comment]
@arelaxend 
I totally appreciate your great work.

However, with my comments I am seeking your help to understand these changes in the PR:
1. Need documentation/reference of your code style.
2. How the grouping works. By optimizing, are you trying to maximize the throughput without hitting the same host too many times? (Sparkler has to offer fair crawling, that it doesnt overload the web servers even if we have large cluster)
3. What was the reason behind taking out the Jbrowser/ Javascript rendering and what are you achieving with Jsoup. How does it work when we have non HTML/XML resources like PDFs/Office-Docs etc

File: conf/solr/crawldb/conf/managed-schema [hunk 2]
```diff
@@ -114,25 +114,32 @@
...
 
    <!-- Crawl DB Static Fields -->
-   <field name="id" type="string" indexed="true" stored="true" required="true" multiValued="false" docValues="true" /> 
+   <field name="id" type="string" indexed="true" stored="true" required="true" multiValued="false" docValues="true" />
    <field name="url" type="string" indexed="true" stored="true" required="true" multiValued="false" docValues="true" />
    <field name="jobId" type="string" indexed="true" stored="true" required="true" multiValued="false" docValues="true" />
-   <field name="group" type="string" indexed="true" stored="true" required="true"  multiValued="false" docValues="true" /> 
-   <field name="lastFetchedAt" type="date" indexed="true" stored="true" multiValued="false" docValues="true" />
+    <field name="group" type="string" indexed="true" stored="true" required="true"  multiValued="false" docValues="true" />
+    <field name="group_id" type="string" indexed="true" stored="true" required="true"  multiValued="false" docValues="true" />
```
Comment (arelaxend) [inline conf/solr/crawldb/conf/managed-schema]: Group is a very good information. For the query, grouping by "group" leads to a non equiprobable distribution in term of size of groups. So I added a new field in which I populate a number which is equally distributed. I used a random generator to fill that "group_id".

File: sparkler-api/src/main/java/edu/usc/irds/sparkler/model/FetchedData.java [hunk 3]
```diff
@@ -17,75 +17,85 @@
...
+    private Resource _resource;
+    private byte[] _content;
+    private String _contentType;
+    private Integer _contentLength;
+    private String[] _headers;
+    private Date _fetchedAt;
+    private int _responseCode;
+    private String[] _outlinks;
+    private String _html;
+    private int _lastOutlinkInserted;
```
Comment (arelaxend) [inline sparkler-api/src/main/java/edu/usc/irds/sparkler/model/FetchedData.java]: When I used "scala friendly", I was not talking about coding conventions. And indeed, you can strongly disagree with this name convention which is not influenced by any open source style of coding convention. Could you provide me some references ?

File: sparkler-api/src/main/java/edu/usc/irds/sparkler/model/Resource.java [hunk 4]
```diff
@@ -31,19 +35,30 @@ public Resource() {
...
-    public Resource(String url, Integer depth, JobContext sparklerJob, ResourceStatus status) throws MalformedURLException {
-        this(url, new URL(url).getHost(), sparklerJob);
+    public Resource(String url, Integer depth, JobContext sparklerJob, ResourceStatus status)
+        throws MalformedURLException {
+        this(url, "", sparklerJob);
+
+        try {
+            setGroup(new URL(url).getHost());
+        } catch(MalformedURLException e) {
+            this.url = "www.error.com";
```
Comment (arelaxend) [inline sparkler-api/src/main/java/edu/usc/irds/sparkler/model/Resource.java]: I was a quick fix, thank you for noticing it!

File: sparkler-api/src/main/java/edu/usc/irds/sparkler/model/Resource.java [hunk 5]
```diff
@@ -58,42 +73,53 @@ public Resource(String url, String group, JobContext sparklerJob, Date lastFetch
...
+            group_id, lastFetchedAt, numTries, numFetches, depth, score, status);
     }
 
 
     public static String resourceId(String url, JobContext job) {
         return String.format("%s-%s", job.getId(), url);
     }
 
+    public void setGroupId() {
+        group_id = String.valueOf(((new Random()).nextInt(GROUP_RANGE) + 1));
```
Comment (arelaxend) [inline sparkler-api/src/main/java/edu/usc/irds/sparkler/model/Resource.java]: Please see my first comment, it's an simple way of having a "equiprobable" distribution between (1, GROUP_RANGE).

File: sparkler-app/pom.xml [hunk 6]
```diff
@@ -121,8 +121,8 @@
 
         <!-- JBrowser Plugin -->
         <dependency>
-            <groupId>com.machinepublishers</groupId>
-            <artifactId>jbrowserdriver</artifactId>
+            <groupId>org.jsoup</groupId>
+            <artifactId>jsoup</artifactId>
             <version>${jbrowserdriver.version}</version>
```
Comment (arelaxend) [inline sparkler-app/pom.xml]: Good that you noticed!

File: conf/felix-config.properties [hunk 7]
```diff
@@ -103,4 +103,4 @@ felix.log.level=1
 org.osgi.framework.system.packages.extra=edu.usc.irds.sparkler; version=0.1,edu.usc.irds.sparkler.util; version=0.1,\
   edu.usc.irds.sparkler.model; version=0.1,org.slf4j; version=1.7.12,org.apache.hadoop.conf; version=2.2.0, \
   javax.net.ssl; version=1.8,com.sun.webkit.network; version=1.8,com.sun.webkit.network.data; version=1.8,\
-  net.jcip.annotations; version=1.8,net.sf.cglib.proxy; version=1.8,com.machinepublishers.jbrowserdriver; version=0.16.4
\ No newline at end of file
+  net.jcip.annotations; version=1.8,net.sf.cglib.proxy; version=1.8,org.jsoup.jsoup; version=1.10.1
```
Comment (arelaxend) [inline conf/felix-config.properties]: It's good that you've noticed! 👍

File: sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/Crawler.scala [hunk 8]
```diff
@@ -88,14 +82,21 @@ class Crawler extends CliTool {
...
 
   def init(): Unit = {
     if (this.outputPath.isEmpty) {
       this.outputPath = jobId
     }
+
+    //http://techkites.blogspot.fr/
     val conf = new SparkConf().setAppName(jobId)
+      //.set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
+
```
Comment (arelaxend) [inline sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/Crawler.scala]: It's good that you've noticed! 👍

File: sparkler-app/src/main/scala/edu/usc/irds/sparkler/service/SolrProxy.scala (hunk 1 above)
Comment (arelaxend) [inline sparkler-app/src/main/scala/edu/usc/irds/sparkler/service/SolrProxy.scala]: It was indeed a better strategy to do batch insertion, and I should be able to make it again with the changes! 👍

File: sparkler-api/src/main/java/edu/usc/irds/sparkler/util/FetcherDefault.java [hunk 9]
```diff
@@ -53,14 +53,18 @@ public FetchedData fetch(Resource resource) throws Exception {
...
-            LOG.warn("FETCH-ERROR {}", resource.getUrl());
+        }
+        catch (Exception e) {
+            LOG.warn("FETCH-ERROR {}", resource.url());
             LOG.debug(e.getMessage(), e);
-            FetchedData fetchedData = new FetchedData(new byte[0], "", DEFAULT_ERROR_CODE);
-            resource.setStatus(ResourceStatus.ERROR.toString());
-            fetchedData.setResource(resource);
-            return fetchedData;
+            return this.fakeFetchedData(resource, ResourceStatus.ERROR);
```
Comment (arelaxend) [inline sparkler-api/src/main/java/edu/usc/irds/sparkler/util/FetcherDefault.java]: 👍

File: sparkler-api/src/main/java/edu/usc/irds/sparkler/util/FetcherDefault.java [hunk 10]
```diff
@@ -53,14 +53,18 @@ public FetchedData fetch(Resource resource) throws Exception {
     public FetchedData apply(Resource resource) {
         try {
             return this.fetch(resource);
-        } catch (Exception e) {
-            LOG.warn("FETCH-ERROR {}", resource.getUrl());
+        }
+        catch (Exception e) {
```
Comment (arelaxend) [inline sparkler-api/src/main/java/edu/usc/irds/sparkler/util/FetcherDefault.java]: 👍

**Comment by arelaxend** [comment]
Thank you for the review @thammegowda ! 👍  🥇 

You noticed good points! At the beginning of the review I said that I removed some parts for the purpose of starting with a fresh version 👍  This does not mean to remove some part forever. That's why I totally agree with you that in 3. a new module must be created separately to handle other ressources but we should keep the proposed one as it works efficiently for HTML files. 

As for the code style, we should keep with what is used in the open source community and I have to improve on this, so thank you for the doc. in ref. and I will look at it.

**Comment by thammegowda** [comment]
> As for the code style, we should keep with what is used in the open source community.

Great :+1: 

> So, what is the correct process of doing ? We close the pull request, I make a new commit and repull ?
1. Close this PR
1. Create issues and describe your proposed changes/suggestions. Hear feedback of other members in team. No feeback == silence == all good, you may proceed
1. Raise new pull requests. We are glad to have many smaller PRs than the one large PR

**Comment by arelaxend** [comment]
(Of course fairness is important for deontological aspects! The idea behind this PR was more about proposing a commit that is fast. Then, to start including more advanced features such as fairness 👍  as a parallel branch and to merge whenever it suits everybody.)

Let's do this process 👍

**Pull request closed without merging**